File: packages/examples/vercel/test-client.ts
#!/usr/bin/env bun

/**
 * Test client for tokagentOS Vercel Edge Functions
 *
 * Usage:
 *   bun run test-client.ts                              # Test local dev server
 *   bun run test-client.ts --endpoint https://your-app.vercel.app  # Test deployed
 *
 * Environment:
 *   VERCEL_URL - Base URL for the Vercel deployment
 */

import * as readline from "node:readline";
import { parseArgs } from "node:util";

// Types
interface ChatResponse {
  response: string;
  conversationId: string;
  timestamp: string;
}

interface HealthResponse {
  status: string;
  runtime: string;
  version: string;
}

interface ErrorResponse {
  error: string;
  code: string;
}

// Parse command line arguments
const { values: args } = parseArgs({
  options: {
    endpoint: { type: "string", short: "e" },
    help: { type: "boolean", short: "h" },
    interactive: { type: "boolean", short: "i" },
  },
});

const DEFAULT_ENDPOINT = "http://localhost:3000";

function showHelp(): void {
  console.log(`
tokagentOS Vercel Edge Function Test Client

Usage:
  bun run test-client.ts [options]

Options:
  -e, --endpoint <url>   API endpoint URL (default: ${DEFAULT_ENDPOINT})
  -i, --interactive      Start interactive chat mode
  -h, --help             Show this help message

Examples:
  bun run test-client.ts                                    # Run tests against local
  bun run test-client.ts -e https://your-app.vercel.app     # Test deployed app
  bun run test-client.ts -i                                 # Interactive chat mode
`);
}

if (args.help) {
  showHelp();
  process.exit(0);
}

const baseUrl = args.endpoint ?? process.env.VERCEL_URL ?? DEFAULT_ENDPOINT;

console.log(`🔗 Using endpoint: ${baseUrl}\n`);

/**
 * Make an HTTP request to the API
 */
async function apiRequest<T>(
  path: string,
  method: string = "GET",
  body?: Record<string, unknown>,
): Promise<{ status: number; data: T }> {
  const url = `${baseUrl}${path}`;
  const options: RequestInit = {
    method,
    headers: {
      "Content-Type": "application/json",
    },
  };

  if (body) {
    options.body = JSON.stringify(body);
  }

  const response = await fetch(url, options);
  const data = (await response.json()) as T;

  return { status: response.status, data };
}

/**
 * Check if the server is available
 */
async function isServerAvailable(): Promise<boolean> {
  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 2000);
    await fetch(`${baseUrl}/api/health`, { signal: controller.signal });
    clearTimeout(timeoutId);
    return true;
  } catch {
    return false;
  }
}

/**
 * Run automated tests
 */
async function runTests(): Promise<void> {
  console.log("🧪 Testing tokagentOS Vercel Edge Functions\n");

  // Check if server is available
  const serverAvailable = await isServerAvailable();
  if (!serverAvailable) {
    console.log(`⚠️  Server not available at ${baseUrl}`);
    console.log("   Skipping integration tests (server must be running)");
    console.log("\n   To run these tests:");
    console.log("   1. Start the server: bun run dev");
    console.log("   2. Run tests: bun run test\n");
    console.log("✅ Tests skipped (no server running)\n");
    process.exit(0);
  }

  let passed = 0;
  let failed = 0;

  // Test 1: Health check
  console.log("1️⃣  Testing health check...");
  try {
    const { status, data } = await apiRequest<HealthResponse>("/api/health");
    console.log(`   Status: ${status}`);
    console.log(`   Runtime: ${data.runtime}`);
    console.log(`   Version: ${data.version}`);

    if (status === 200 && data.status === "healthy") {
      console.log("   ✅ Health check passed\n");
      passed++;
    } else {
      console.log("   ❌ Health check failed\n");
      failed++;
    }
  } catch (error) {
    console.log(
      `   ❌ Error: ${error instanceof Error ? error.message : "Unknown"}\n`,
    );
    failed++;
  }

  // Test 2: Chat endpoint
  console.log("2️⃣  Testing chat endpoint...");
  try {
    const start = Date.now();
    const { status, data } = await apiRequest<ChatResponse>(
      "/api/chat",
      "POST",
      {
        message: "Hello! What's 2 + 2?",
      },
    );
    const duration = Date.now() - start;

    console.log(`   Status: ${status}`);
    console.log(`   Duration: ${duration}ms`);
    console.log(`   Conversation ID: ${data.conversationId}`);
    console.log(
      `   Response: ${data.response.slice(0, 100)}${data.response.length > 100 ? "..." : ""}`,
    );

    if (status === 200 && data.response) {
      console.log("   ✅ Chat endpoint passed\n");
      passed++;
    } else {
      console.log("   ❌ Chat endpoint failed\n");
      failed++;
    }
  } catch (error) {
    console.log(
      `   ❌ Error: ${error instanceof Error ? error.message : "Unknown"}\n`,
    );
    failed++;
  }

  // Test 3: Validation (empty message)
  console.log("3️⃣  Testing validation (empty message)...");
  try {
    const { status, data } = await apiRequest<ErrorResponse>(
      "/api/chat",
      "POST",
      {
        message: "",
      },
    );
    console.log(`   Status: ${status}`);
    console.log(`   Error: ${data.error}`);

    if (status === 400 && data.code === "BAD_REQUEST") {
      console.log("   ✅ Validation passed\n");
      passed++;
    } else {
      console.log("   ❌ Validation failed\n");
      failed++;
    }
  } catch (error) {
    console.log(
      `   ❌ Error: ${error instanceof Error ? error.message : "Unknown"}\n`,
    );
    failed++;
  }

  // Test 4: 404 handling
  console.log("4️⃣  Testing 404 response...");
  try {
    const { status, data } = await apiRequest<ErrorResponse>("/api/unknown");
    console.log(`   Status: ${status}`);

    if (status === 404 && data.code === "NOT_FOUND") {
      console.log("   ✅ 404 handling passed\n");
      passed++;
    } else {
      console.log("   ❌ 404 handling failed\n");
      failed++;
    }
  } catch (error) {
    console.log(
      `   ❌ Error: ${error instanceof Error ? error.message : "Unknown"}\n`,
    );
    failed++;
  }

  // Test 5: Method not allowed
  console.log("5️⃣  Testing method not allowed...");
  try {
    const { status, data } = await apiRequest<ErrorResponse>(
      "/api/chat",
      "GET",
    );
    console.log(`   Status: ${status}`);

    if (status === 405 && data.code === "METHOD_NOT_ALLOWED") {
      console.log("   ✅ Method handling passed\n");
      passed++;
    } else {
      console.log("   ❌ Method handling failed\n");
      failed++;
    }
  } catch (error) {
    console.log(
      `   ❌ Error: ${error instanceof Error ? error.message : "Unknown"}\n`,
    );
    failed++;
  }

  // Summary
  console.log("━".repeat(40));
  console.log(`Results: ${passed} passed, ${failed} failed`);

  if (failed > 0) {
    console.log("\n❌ Some tests failed!");
    process.exit(1);
  } else {
    console.log("\n🎉 All tests passed!");
  }
}

/**
 * Interactive chat mode
 */
async function interactiveMode(): Promise<void> {
  console.log("💬 Interactive Chat Mode");
  console.log('   Type your message and press Enter. Type "exit" to quit.\n');

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  let conversationId: string | undefined;

  const prompt = (): void => {
    rl.question("You: ", async (input: string) => {
      const message = input.trim();

      if (message.toLowerCase() === "exit") {
        console.log("\n👋 Goodbye!");
        rl.close();
        return;
      }

      if (!message) {
        prompt();
        return;
      }

      const { data } = await apiRequest<ChatResponse>("/api/chat", "POST", {
        message,
        conversationId,
      });

      conversationId = data.conversationId;
      console.log(`\nTokagent: ${data.response}\n`);

      prompt();
    });
  };

  prompt();
}

// Main
if (args.interactive) {
  interactiveMode();
} else {
  runTests();
}

File: packages/examples/vercel/test-client.py
#!/usr/bin/env python3
"""
Test client for elizaOS Vercel Edge Functions

Usage:
    python3 test-client.py                                    # Test local
    python3 test-client.py --endpoint https://your-app.vercel.app  # Test deployed

Environment:
    VERCEL_URL - Base URL for the Vercel deployment
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.request
from typing import TypedDict


class ChatResponse(TypedDict):
    response: str
    conversationId: str
    timestamp: str


class HealthResponse(TypedDict):
    status: str
    runtime: str
    version: str


class ErrorResponse(TypedDict):
    error: str
    code: str


DEFAULT_ENDPOINT = "http://localhost:3000"


def api_request(base_url: str, path: str, method: str = "GET", body: dict | None = None) -> tuple[int, dict]:
    """Make an HTTP request to the API."""
    url = f"{base_url}{path}"
    data = json.dumps(body).encode() if body else None
    headers = {"Content-Type": "application/json"}

    req = urllib.request.Request(url, data=data, headers=headers, method=method)

    try:
        with urllib.request.urlopen(req, timeout=60) as response:
            return response.status, json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read().decode())
    except Exception as e:
        return 0, {"error": str(e), "code": "CONNECTION_ERROR"}


def run_tests(base_url: str) -> bool:
    """Run automated tests."""
    print("🧪 Testing elizaOS Vercel Edge Functions\n")

    passed = 0
    failed = 0

    # Test 1: Health check
    print("1️⃣  Testing health check...")
    status, data = api_request(base_url, "/api/health")
    print(f"   Status: {status}")
    print(f"   Runtime: {data.get('runtime', 'N/A')}")
    print(f"   Version: {data.get('version', 'N/A')}")

    if status == 200 and data.get("status") == "healthy":
        print("   ✅ Health check passed\n")
        passed += 1
    else:
        print("   ❌ Health check failed\n")
        failed += 1

    # Test 2: Chat endpoint
    print("2️⃣  Testing chat endpoint...")
    start = time.time()
    status, data = api_request(base_url, "/api/chat", "POST", {
        "message": "Hello! What's 2 + 2?"
    })
    duration = int((time.time() - start) * 1000)

    print(f"   Status: {status}")
    print(f"   Duration: {duration}ms")
    print(f"   Conversation ID: {data.get('conversationId', 'N/A')}")
    response_text = data.get("response", "")
    print(f"   Response: {response_text[:100]}{'...' if len(response_text) > 100 else ''}")

    if status == 200 and response_text:
        print("   ✅ Chat endpoint passed\n")
        passed += 1
    else:
        print("   ❌ Chat endpoint failed\n")
        failed += 1

    # Test 3: Validation (empty message)
    print("3️⃣  Testing validation (empty message)...")
    status, data = api_request(base_url, "/api/chat", "POST", {"message": ""})
    print(f"   Status: {status}")
    print(f"   Error: {data.get('error', 'N/A')}")

    if status == 400 and data.get("code") == "BAD_REQUEST":
        print("   ✅ Validation passed\n")
        passed += 1
    else:
        print("   ❌ Validation failed\n")
        failed += 1

    # Test 4: 404 handling
    print("4️⃣  Testing 404 response...")
    status, data = api_request(base_url, "/api/unknown")
    print(f"   Status: {status}")

    if status == 404 and data.get("code") == "NOT_FOUND":
        print("   ✅ 404 handling passed\n")
        passed += 1
    else:
        print("   ❌ 404 handling failed\n")
        failed += 1

    # Test 5: Method not allowed
    print("5️⃣  Testing method not allowed...")
    status, data = api_request(base_url, "/api/chat", "GET")
    print(f"   Status: {status}")

    if status == 405 and data.get("code") == "METHOD_NOT_ALLOWED":
        print("   ✅ Method handling passed\n")
        passed += 1
    else:
        print("   ❌ Method handling failed\n")
        failed += 1

    # Summary
    print("━" * 40)
    print(f"Results: {passed} passed, {failed} failed")

    if failed > 0:
        print("\n❌ Some tests failed!")
        return False
    else:
        print("\n🎉 All tests passed!")
        return True


def interactive_mode(base_url: str) -> None:
    """Run interactive chat mode."""
    print("💬 Interactive Chat Mode")
    print('   Type your message and press Enter. Type "exit" to quit.\n')

    conversation_id: str | None = None

    while True:
        try:
            message = input("You: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\n👋 Goodbye!")
            break

        if message.lower() == "exit":
            print("\n👋 Goodbye!")
            break

        if not message:
            continue

        body = {"message": message}
        if conversation_id:
            body["conversationId"] = conversation_id

        status, data = api_request(base_url, "/api/chat", "POST", body)

        if status == 200:
            conversation_id = data.get("conversationId")
            print(f"\nEliza: {data.get('response', 'No response')}\n")
        else:
            print(f"\n❌ Error: {data.get('error', 'Unknown error')}\n")


def main() -> None:
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="elizaOS Vercel Edge Function Test Client"
    )
    parser.add_argument(
        "-e", "--endpoint",
        default=os.environ.get("VERCEL_URL", DEFAULT_ENDPOINT),
        help=f"API endpoint URL (default: {DEFAULT_ENDPOINT})"
    )
    parser.add_argument(
        "-i", "--interactive",
        action="store_true",
        help="Start interactive chat mode"
    )

    args = parser.parse_args()

    print(f"🔗 Using endpoint: {args.endpoint}\n")

    if args.interactive:
        interactive_mode(args.endpoint)
    else:
        success = run_tests(args.endpoint)
        sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()










